File: common/utils/sw/dukpt.py
"""
AES DUKPT (Derived Unique Key Per Transaction) Implementation

This implementation is based on the reference implementation provided by the
Accredited Standards Committee X9 (ASC X9) for ANSI X9.24-3-2017.

Reference Implementation Source:
- X9.24 Part 3 Test Vectors: https://x9.org/standards/x9-24-part-3-test-vectors/
- Python Source Code: https://x9.org/wp-content/uploads/2018/03/X9.24-3-2017-Python-Source-20180129-1.pdf

Original X9 Disclaimer:
"The information is provided 'as is' without warranty of any kind. X9 does not accept 
any responsibility or liability for the accuracy, content, completeness, or reliability 
of the computer code and information contained on this page and website."

This implementation has been modified from the original X9 reference implementation
to integrate with the fmcrypto Service architecture and requirements.

Copyright (c) 2025 fmcrypto Service
"""

from Crypto.Cipher import AES
from enum import Enum

Debug = False
NUMREG = 32
MAX_WORK = 16

# convert a 32-bit integer to a list of bytes in big-endian order.  Used to convert counter values 
# to byte lists.
def int_to_bytes(x):
    return [((x >> i) & 0xff) for i in (24,16,8,0)]


# B.3.1. enumerations
class DerivationPurpose(Enum):
    _InitialKey = 0
    _DerivationOrWorkingKey = 1

class KeyType(Enum):
    _2TDEA = 0
    _3TDEA = 1
    _AES128 = 2
    _AES192 = 3
    _AES256 = 4

class KeyUsage(Enum):
    _KeyEncryptionKey = 0x0002
    _PINEncryption = 0x1000
    _MessageAuthenticationGeneration = 0x2000
    _MessageAuthenticationVerification = 0x2001
    _MessageAuthenticationBothWays = 0x2002
    _DataEncryptionEncrypt = 0x3000
    _DataEncryptionDecrypt = 0x3001
    _DataEncryptionBothWays = 0x3002
    _KeyDerivation = 0x8000
    _KeyDerivationInitialKey = 9

# count the number of 1 bits in a counter value.  readable, but not efficient.
def count_one_bits(x):
    bits = 0
    mask = 1 << (NUMREG-1)
    while mask > 0:
        if x & mask:
            bits = bits + 1
        mask = mask >> 1
    return bits

# B.3.2. key length function
# length of an algorithm's key, in bits.
def key_length(key_type):
    if ( key_type == KeyType._2TDEA):
        return 128
    if ( key_type == KeyType._3TDEA):
        return 192
    if ( key_type == KeyType._AES128):
        return 128
    if ( key_type == KeyType._AES192):
        return 192
    if ( key_type == KeyType._AES256):
        return 256
    assert False

# encrypt plaintext with key using AES.
def aes_encrypt_ecb(key, plaintext):
    # ECB mode is specifically required by the ANSI X9.24-3-2017 standard for DUKPT key derivation
    # this is not a security vulnerability in this context as it's used for key derivation - 
    # not data encryption
    # nosonar python:S5542 - AES ECB required for X9.24-3 standard compliance
    encobj = AES.new(bytes(key), AES.MODE_ECB) 
    result = encobj.encrypt(bytes(plaintext))
    return result

# compute the xor of two 128-bit numbers
def xor(a, b):
    result = bytearray(16)
    for i in range(0,16):
        result[i] = a[i] ^ b[i]
    return result

# B.4.1. derive key algorithm
# AES DUKPT key derivation function.    
def derive_key(derivation_key, key_type, derivation_data, _):
    L = key_length(key_type)
    n = -(-L // 128)

    result = bytearray(n*16)
    for i in range(1, n+1):
        derivation_data[1] = i
        result[(i-1)*16:i*16] = aes_encrypt_ecb(derivation_key, derivation_data)

    derived_key = result[0:(L//8)]
    return derived_key

def _set_key_usage(key_usage, derivation_data):
    if (key_usage == KeyUsage._KeyEncryptionKey):
        derivation_data[2:4] = [0,2]
    elif (key_usage == KeyUsage._PINEncryption):
        derivation_data[2:4] = [16,0]
    elif (key_usage == KeyUsage._MessageAuthenticationGeneration):
        derivation_data[2:4] = [32,0]
    elif (key_usage == KeyUsage._MessageAuthenticationVerification):
        derivation_data[2:4] = [32,1]
    elif (key_usage == KeyUsage._MessageAuthenticationBothWays):
        derivation_data[2:4] = [32,2]
    elif (key_usage == KeyUsage._DataEncryptionEncrypt):
        derivation_data[2:4] = [48,0]
    elif (key_usage == KeyUsage._DataEncryptionDecrypt):
        derivation_data[2:4] = [48,1]
    elif (key_usage == KeyUsage._DataEncryptionBothWays):
        derivation_data[2:4] = [48,2]
    elif (key_usage == KeyUsage._KeyDerivation):
        derivation_data[2:4] = [128,0]
    elif (key_usage == KeyUsage._KeyDerivationInitialKey):
        derivation_data[2:4] = [128,1]
    else:
        assert False

def _set_derived_key_type(derived_key_type, derivation_data):
    if (derived_key_type == KeyType._2TDEA):
        derivation_data[4:6] = [0,0]
    elif (derived_key_type == KeyType._3TDEA):
        derivation_data[4:6] = [0,1]
    elif (derived_key_type == KeyType._AES128):
        derivation_data[4:6] = [0,2]
    elif (derived_key_type == KeyType._AES192):
        derivation_data[4:6] = [0,3]
    elif (derived_key_type == KeyType._AES256):
        derivation_data[4:6] = [0,4]
    else:
        assert False


# B.4.3. create derivation data
# compute derivation data for an AES DUKPT key derivation operation.
def create_derivation_data(
    derivation_purpose, key_usage, derived_key_type, initial_key_id, counter):
    derivation_data = bytearray(16)
    derivation_data[0] = 1
    derivation_data[1] = 1
    # set key usage
    _set_key_usage(key_usage, derivation_data)
    # set derived key type
    _set_derived_key_type(derived_key_type, derivation_data)
    # set derivation purpose
    if (derivation_purpose == DerivationPurpose._InitialKey):
        derivation_data[8:16] = initial_key_id[0:8]
    elif (derivation_purpose == DerivationPurpose._DerivationOrWorkingKey):
        derivation_data[8:12] = initial_key_id[4:8]
        derivation_data[12:16] = int_to_bytes(counter)
    else:
        assert False
    return derivation_data


# B.6.3. processing routines
# load an initial key for computing terminal transaction keys in sequence.
def load_initial_key(initial_key, derive_key_type, initial_key_id):
    global NUMREG
    global g_intermediate_derivation_key_register
    global g_intermediate_derivation_key_in_use
    global g_current_key
    global g_device_id
    global g_counter
    global g_shift_register
    global g_derive_key_type
    # set intermediate derivation key register
    g_intermediate_derivation_key_register = [None]*NUMREG
    g_intermediate_derivation_key_in_use = [False]*NUMREG
    g_intermediate_derivation_key_register[0] = initial_key
    g_intermediate_derivation_key_in_use[0] = True
    g_device_id = initial_key_id
    g_counter = 0
    g_shift_register = 1
    g_current_key = 0
    g_derive_key_type = derive_key_type
    # update derivation keys
    update_derivation_keys(NUMREG-1, derive_key_type)
    g_counter = g_counter + 1


# B.6.3. generate working keys
# generate a transaction key from the intermediate derivation key registers, and update the state 
# to prepare for the next transaction.
def generate_working_keys(working_key_usage, working_key_type):
    global NUMREG
    global g_intermediate_derivation_key_register
    global g_intermediate_derivation_key_in_use
    global g_current_key
    global g_device_id
    global g_counter
    global g_shift_register
    global g_derive_key_type
    # set shift register
    set_shift_register()

    while not g_intermediate_derivation_key_in_use[g_current_key]:
        g_counter = g_counter + g_shift_register
        if g_counter > ((1 << NUMREG) - 1):
            # Counter overflow - cease operation as per X9.24-3 standard
            return False
        set_shift_register()
    # create derivation data
    derivation_data = create_derivation_data(
        DerivationPurpose._DerivationOrWorkingKey, working_key_usage, working_key_type, 
        g_device_id, g_counter)
    assert g_intermediate_derivation_key_in_use[g_current_key]
    working_key = derive_key(
        g_intermediate_derivation_key_register[g_current_key], working_key_type, derivation_data, 
        g_derive_key_type)
    
    update_state_for_next_transactio ()
    return working_key

# B.6.3. update state for next transaction
# move the counter forward, and derive new intermediate derivation keys for the next transaction.
def update_state_for_next_transactio ():
    global NUMREG
    global MAX_WORK
    global g_intermediate_derivation_key_register
    global g_intermediate_derivation_key_in_use
    global g_current_key
    global g_device_id
    global g_counter
    global g_shift_register
    global g_derive_key_type
    
    one_bits  = count_one_bits(g_counter)
    if one_bits  <= MAX_WORK:
        update_derivation_keys(g_current_key, g_derive_key_type)
        g_intermediate_derivation_key_register[g_current_key] = 0
        g_intermediate_derivation_key_in_use[g_current_key] = False
        g_counter = g_counter + 1
    else:
        g_intermediate_derivation_key_register[g_current_key] = 0
        g_intermediate_derivation_key_in_use[g_current_key] = False
        g_counter = g_counter + g_shift_register

    if g_counter > (1 << NUMREG) - 1:
        # Counter overflow - cease operation as per X9.24-3 standard
        return False
    else:
        return True

# B.6.3. update derivation keys
# update all the intermediate derivation key registers below a certain point.
# this is used to:
# 1. update all the intermediate derivation key registers below the shift register after computing 
# a transaction key.
# 2. update all the intermediate derivation key registers when an initial key is loaded.
def update_derivation_keys(start, derive_key_type):
    global NUMREG
    global g_intermediate_derivation_key_register
    global g_intermediate_derivation_key_in_use
    global g_current_key
    global g_device_id
    global g_counter
    global g_shift_register

    i = start
    j = 1 << start

    base_key = g_intermediate_derivation_key_register[g_current_key]
    while j != 0:
        derivation_data = create_derivation_data(
            DerivationPurpose._DerivationOrWorkingKey, KeyUsage._KeyDerivation, derive_key_type, 
            g_device_id, g_counter | j)
        assert g_intermediate_derivation_key_in_use[g_current_key]
        g_intermediate_derivation_key_register[i] = derive_key(
            base_key, derive_key_type, derivation_data, derive_key_type)
        g_intermediate_derivation_key_in_use[i] = True
        j = j >> 1
        i = i - 1
    return True

# B.6.3. set shift register
# set the shift register to the value of the rightmost '1' bit in the counter.
def set_shift_register():
    global NUMREG
    global g_intermediate_derivation_key_register
    global g_intermediate_derivation_key_in_use
    global g_current_key
    global g_device_id
    global g_counter
    global g_shift_register

    g_shift_register = 1
    g_current_key = 0

    ret = True
    if g_counter == 0:
        return ret

    while (g_shift_register & g_counter) == 0:
        g_shift_register = g_shift_register << 1
        g_current_key = g_current_key + 1

    return True

class Dukpt:
    # B.5. derive initial key
    # derive the initial key for a particular initial_key_id from a BDK.
    def derive_initial_key(self, bdk, key_type, initial_key_id):
        derivation_data = create_derivation_data(
            DerivationPurpose._InitialKey, KeyUsage._KeyDerivationInitialKey, key_type, 
            initial_key_id, 0)
        initial_key = derive_key(bdk, key_type, derivation_data, key_type)
        return initial_key

    def derive_working_key(
        self, initial_key, derive_key_type, working_key_usage, working_key_type, initial_key_id, 
        counter):
        mask = 0x80000000
        working_counter = 0
        derivation_key = initial_key

        while mask > 0:
            if (mask & counter) != 0:
                working_counter = working_counter | mask
                derivation_data = create_derivation_data(DerivationPurpose._DerivationOrWorkingKey, KeyUsage._KeyDerivation, derive_key_type, initial_key_id, working_counter)
                derivation_key = derive_key(derivation_key, derive_key_type, derivation_data, derive_key_type)
            mask = mask >> 1

        derivation_data = create_derivation_data(DerivationPurpose._DerivationOrWorkingKey, working_key_usage, working_key_type, initial_key_id, counter)
        working_key = derive_key(derivation_key, working_key_type, derivation_data, derive_key_type)

        return derivation_key, derivation_data, working_key

    # B.5. host derive working key
    # derive a working key for a particular transaction based on a initial_key_id and counter.
    def host_derive_working_key(
        self, bdk, derive_key_type, working_key_usage, working_key_type, initial_key_id, counter):
        initial_key = self.derive_initial_key(bdk, derive_key_type, initial_key_id)
        print(f"intial key: {initial_key.hex()}")
        
        derivation_key, derivation_data, working_key = self.derive_working_key(
            initial_key, derive_key_type, working_key_usage, working_key_type, initial_key_id, 
            counter)
        return derivation_key, derivation_data, working_key
    

File: services/crypto_svc/internal/crypto/dukpt/dukpt.go
// Package dukpt implements host-side AES DUKPT (Derived Unique Key Per
// Transaction) key derivation per ANSI X9.24-3-2017.
//
// This implementation is based on the reference implementation provided by the
// Accredited Standards Committee X9 (ASC X9) for ANSI X9.24-3-2017.
//
// Reference Implementation Source:
// - X9.24 Part 3 Test Vectors: https://x9.org/standards/x9-24-part-3-test-vectors/
// - Python Source Code: https://x9.org/wp-content/uploads/2018/03/X9.24-3-2017-Python-Source-20180129-1.pdf
//
// Original X9 Disclaimer:
// "The information is provided 'as is' without warranty of any kind. X9 does
// not accept any responsibility or liability for the accuracy, content,
// completeness, or reliability of the computer code and information contained
// on this page and website."
//
// This implementation has been modified from the original X9 reference
// implementation to integrate with the fmcrypto Service architecture and
// requirements.
//
// Copyright (c) 2025 fmcrypto Service
package dukpt

import (
	"crypto/aes"
	"fmt"
)

// KeyType enumerates the supported DUKPT key algorithms (X9.24-3 §B.3.1).
type KeyType int

const (
	TwoTDEA   KeyType = 0
	ThreeTDEA KeyType = 1
	AES128    KeyType = 2
	AES192    KeyType = 3
	AES256    KeyType = 4
)

// KeyUsage enumerates the derived-key purposes (X9.24-3 §B.3.1).
type KeyUsage int

const (
	KeyEncryptionKey      KeyUsage = 0x0002
	PINEncryption         KeyUsage = 0x1000
	MessageAuthGen        KeyUsage = 0x2000
	MessageAuthVerify     KeyUsage = 0x2001
	MessageAuthBothWays   KeyUsage = 0x2002
	DataEncryptionEncrypt KeyUsage = 0x3000
	DataEncryptionDecrypt KeyUsage = 0x3001
	DataEncryptionBoth    KeyUsage = 0x3002
	KeyDerivation         KeyUsage = 0x8000
	KeyDerivationInitial  KeyUsage = 9
)

type derivationPurpose int

const (
	purposeInitialKey derivationPurpose = 0
	purposeWorkingKey derivationPurpose = 1
)

// Deriver derives initial and transaction keys from a BDK.
type Deriver struct{}

// DeriveInitialKey derives the initial key for an initial-key ID from a BDK.
func (Deriver) DeriveInitialKey(bdk []byte, keyType KeyType, initialKeyID []byte) ([]byte, error) {
	d := createDerivationData(purposeInitialKey, KeyDerivationInitial, keyType, initialKeyID, 0)
	return deriveKey(bdk, keyType, d)
}

// DeriveWorkingKey derives the derivation key, derivation data, and working key
// for a transaction identified by counter.
func (Deriver) DeriveWorkingKey(initialKey []byte, deriveKeyType KeyType,
	workingKeyUsage KeyUsage, workingKeyType KeyType,
	initialKeyID []byte, counter uint32) (derivationKey, derivationData, workingKey []byte, err error) {
	var mask uint32 = 0x80000000
	workingCounter := uint32(0)
	derivationKey = initialKey

	for mask > 0 {
		if mask&counter != 0 {
			workingCounter |= mask
			d := createDerivationData(purposeWorkingKey, KeyDerivation, deriveKeyType, initialKeyID, workingCounter)
			dk, err := deriveKey(derivationKey, deriveKeyType, d)
			if err != nil {
				return nil, nil, nil, err
			}
			derivationKey = dk
		}
		mask >>= 1
	}

	d := createDerivationData(purposeWorkingKey, workingKeyUsage, workingKeyType, initialKeyID, counter)
	wk, err := deriveKey(derivationKey, workingKeyType, d)
	if err != nil {
		return nil, nil, nil, err
	}
	return derivationKey, d, wk, nil
}

// HostDeriveWorkingKey derives a transaction working key directly from a BDK.
func (Deriver) HostDeriveWorkingKey(bdk []byte, deriveKeyType KeyType,
	workingKeyUsage KeyUsage, workingKeyType KeyType,
	initialKeyID []byte, counter uint32) ([]byte, []byte, []byte, error) {
	ik, err := (Deriver{}).DeriveInitialKey(bdk, deriveKeyType, initialKeyID)
	if err != nil {
		return nil, nil, nil, err
	}
	return (Deriver{}).DeriveWorkingKey(ik, deriveKeyType, workingKeyUsage, workingKeyType, initialKeyID, counter)
}

// --- internals (X9.24-3 §B.4) ---

func keyLength(t KeyType) int {
	switch t {
	case TwoTDEA:
		return 128
	case ThreeTDEA:
		return 192
	case AES128:
		return 128
	case AES192:
		return 192
	case AES256:
		return 256
	}
	return 0
}

// aesECBEncrypt encrypts a 16-byte block. ECB is used only for key derivation,
// as required by the standard — never for data encryption.
func aesECBEncrypt(key, plaintext []byte) ([]byte, error) {
	if len(key) != 16 && len(key) != 24 && len(key) != 32 {
		return nil, fmt.Errorf("dukpt: invalid key length %d", len(key))
	}
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}
	if len(plaintext) != block.BlockSize() {
		return nil, fmt.Errorf("dukpt: input must be one block")
	}
	out := make([]byte, block.BlockSize())
	block.Encrypt(out, plaintext)
	return out, nil
}

func intToBytes(x uint32) []byte {
	return []byte{byte(x >> 24), byte(x >> 16), byte(x >> 8), byte(x)}
}

func setKeyUsage(u KeyUsage, d []byte) {
	switch u {
	case KeyEncryptionKey:
		d[2], d[3] = 0, 2
	case PINEncryption:
		d[2], d[3] = 16, 0
	case MessageAuthGen:
		d[2], d[3] = 32, 0
	case MessageAuthVerify:
		d[2], d[3] = 32, 1
	case MessageAuthBothWays:
		d[2], d[3] = 32, 2
	case DataEncryptionEncrypt:
		d[2], d[3] = 48, 0
	case DataEncryptionDecrypt:
		d[2], d[3] = 48, 1
	case DataEncryptionBoth:
		d[2], d[3] = 48, 2
	case KeyDerivation:
		d[2], d[3] = 128, 0
	case KeyDerivationInitial:
		d[2], d[3] = 128, 1
	}
}

func setKeyType(t KeyType, d []byte) {
	switch t {
	case TwoTDEA:
		d[4], d[5] = 0, 0
	case ThreeTDEA:
		d[4], d[5] = 0, 1
	case AES128:
		d[4], d[5] = 0, 2
	case AES192:
		d[4], d[5] = 0, 3
	case AES256:
		d[4], d[5] = 0, 4
	}
}

// deriveKey implements the AES DUKPT key derivation function (§B.4.1).
func deriveKey(derivationKey []byte, keyType KeyType, derivationData []byte) ([]byte, error) {
	l := keyLength(keyType)
	n := (l + 127) / 128
	out := make([]byte, 0, n*16)
	for i := 1; i <= n; i++ {
		derivationData[1] = byte(i)
		block, err := aesECBEncrypt(derivationKey, derivationData)
		if err != nil {
			return nil, err
		}
		out = append(out, block...)
	}
	return out[:l/8], nil
}

func createDerivationData(purpose derivationPurpose, usage KeyUsage,
	derivedType KeyType, initialKeyID []byte, counter uint32) []byte {
	d := make([]byte, 16)
	d[0], d[1] = 1, 1
	setKeyUsage(usage, d)
	setKeyType(derivedType, d)
	switch purpose {
	case purposeInitialKey:
		copy(d[8:16], initialKeyID[0:8])
	case purposeWorkingKey:
		copy(d[8:12], initialKeyID[4:8])
		copy(d[12:16], intToBytes(counter))
	}
	return d
}
